038-optimization-proposal-autoapproval.md

```diff
@@ -0,0 +1,99 @@
+It enables more automation where just creating a `KafkaRebalance` custom resource can go straight to the cluster rebalance without any additional manual interaction.
+
+## Proposal
+
+The `KafkaRebalance` custom resource can be extended with a new `spec.autoApproval` field for this purpose.
+The auto-approval can then be enabled or disable through a boolean `spec.autoApproval.enabled` field.
+
+### Auto-approval
+
+The auto-approval can be enabled through the boolean `spec.autoApproval.enabled` field set to `true`.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval:
+    enabled: true
+```
+
+The same result can be easily achieved if the `spec.autoApproval` is specified as empty.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval: {}
+```
+
+The reason for not having a simple boolean `spec.autoApproval` field is for leaving space for a future extension, maybe adding some criteria or rules that have to be matched for auto-approving the optimization proposal.
+
+### No auto-approval (default)
+
+If the `spec.autoApproval` field is not specified at all, the default behavior will be the current one, so the need for manual approval by the user annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
+
+Of course, setting the `spec.autoApproval.enabled` to `false` drives to the same result.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval:
+    enabled: false
+```
+
+### Flow
+
+As described before, the user interaction flow assumes that the `spec.autoApproval` field is specified in the `KafkaRebalance` custom resource.
+
+1. The user creates a `KafkaRebalance` custom resource.
+2. The cluster operator asks Cruise Control to generate an optimization proposal via the REST API.
+3. When the optimization proposal is ready, the cluster operator checks if the user has specified the `spec.autoApproval` and the `spec.autoApproval.enabled` is set or not:
+    * If the `spec.autoApproval` is not specified at all, the user has to approve the proposal manually as usual.
+    * If the `spec.autoApproval` is specified but it's not enabled with `spec.autoApproval.enabled` set to `false`, the user has to approve the proposal manually as usual.
+    * If the `spec.autoApproval` is specified and it's enabled with `spec.autoApproval.enabled` set to `true`, the cluster operator approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
```

Comment:
Also ... you say `the cluster operator approves the proposal automatically by annotating the KafkaRebalance custom resource with the strimzi.io/rebalance=approve annotation`. Will it really set the annotation? Or just approve it inside the code and move it into _rebalancing_ state in the code only?

Reply:
In general, I would always rely on the current state machine (by applying the annotation) instead of having a specific check to run the same code for this scenario.

Reply:
Ok, just wanted to make sure this is what you really wanted to do.